wagtail/users/views/users.py

```diff
@@ -180,57 +182,72 @@ def create(request):
-def edit(request, user_id):
-    user = get_object_or_404(User, pk=user_id)
-    can_delete = user_can_delete_user(request.user, user)
-    editing_self = request.user == user
+class Edit(EditView):
+
+    model = User
+    permission_policy = ModelPermissionPolicy(User)
+    form_class = get_user_edit_form()
+    template_name = "wagtailusers/users/edit.html"
+    index_url_name = "wagtailusers_users:index"
+    success_url = "wagtailusers_users:index"
+    edit_url_name = "wagtailusers_users:edit"
+    delete_url_name = "wagtailusers_users:delete"
+    success_message = _("User '{0}' updated.")
+    context_object_name = "user"
+
+    def get_page_title(self):
+        return "Editing {}".format(self.object.get_username)
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        self.object = self.get_object()
+        self.can_delete = user_can_delete_user(request.user, self.object)
+        self.editing_self = request.user == self.object
+
+    def save_instance(self):
+        instance = super().save_instance()
+        if self.object == self.request.user and "password1" in self.form.changed_data:
+            # User is changing their own password; need to update their session hash
+            update_session_auth_hash(self.request, self.object)
+        return instance
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs.update({"editing_self": self.editing_self})
+        return kwargs
+
+    def run_before_hook(self):
+        for fn in hooks.get_hooks("before_edit_user"):
+            result = fn(self.request, self.object)
+            if hasattr(result, "status_code"):
+                return result
 
-    for fn in hooks.get_hooks("before_edit_user"):
-        result = fn(request, user)
-        if hasattr(result, "status_code"):
-            return result
-    if request.method == "POST":
-        form = get_user_edit_form()(
-            request.POST, request.FILES, instance=user, editing_self=editing_self
+    def run_after_hook(self):
+        for fn in hooks.get_hooks("after_edit_user"):
+            result = fn(self.request, self.object)
+            if hasattr(result, "status_code"):
+                return result
+
+    def get_success_buttons(self):
+        return (
+            [
+                messages.button(
+                    reverse("wagtailusers_users:edit", args=(self.object.pk,)),
+                    _("Edit"),
+                )
+            ],
         )
-        if form.is_valid():
-            with transaction.atomic():
-                user = form.save()
-                log(user, "wagtail.edit")
 
-            if user == request.user and "password1" in form.changed_data:
-                # User is changing their own password; need to update their session hash
-                update_session_auth_hash(request, user)
+    def get_edit_url(self):
+        return reverse(self.edit_url_name, args=(self.object.pk,))
```

Comment:
In our case, `CustomUser` does not have an `id` field. So, we need to override the whole method to access the URL using `pk` instead of `id`. This is also the case for overriding `get_success_buttons`. I think we can alter the base class so we can use an arbitrary field similar to `pk_url_kwarg` in `SingleObjectMixin` in the Django project.

Reply:
I think it would be reasonable to fix `wagtail.admin.views.generic.EditView` to use `object.pk` instead of `object.id`. (We should also ensure we're using `quote` / `unquote` consistently to handle non-url-safe primary keys - it seems IndexView already does this for `get_edit_url` but EditView doesn't unquote it at the other end - but that's a separate issue...)

Reply:
Thanks for the review @gasman. Also, for the `quote` / `unquote`, should I wait for the first one to merge and then jump to that?

Reply:
That's up to you... I'm probably not going to have chance to review this for a few days, and I think the `quote`/`unquote` fix could reasonably go in the same PR.

Reply:
See created issue from this thread - https://github.com/wagtail/wagtail/issues/9205
and PR https://github.com/wagtail/wagtail/pull/9221